### PR TITLE
fix: Fix upgrade path 3.3.x -> 4.0.x

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Fixes https://github.com/stackabletech/hive-operator/issues/538

This PR also adds an upgrade test to make sure the functionality will be kept going forward.

Co-developed with @xeniape

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM

**Comment by NickLarsenNZ** [comment]
Btw, you can add a `Co-authered-by` trailer to the commit to give @xeniape attribution (not sure if `Co-developed` has any significance in github if that's what you were going for).

File: rust/operator-binary/src/controller.rs [hunk 1]
```diff
@@ -828,36 +821,24 @@ fn build_metastore_rolegroup_statefulset(
...
-                            }
-                        })?);
-                    }
+        if property_name_kind == &PropertyNameKind::Env {
+            // overrides
+            for (property_name, property_value) in config {
+                if property_name.is_empty() {
+                    warn!(
+                        property_name,
+                        "Received empty property_name for ENV... skipping"
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/controller.rs]: It was existing code, but it's refering to env variables. I included property_name and property_value in the message in be50ac4 and 2660bae

File: rust/operator-binary/src/controller.rs [hunk 2]
```diff
@@ -894,6 +875,26 @@ fn build_metastore_rolegroup_statefulset(
...
+        // TODO: Once we drop support for HMS 3.1.x we can remove this condition and very likely get rid of the
+        // "bin/start-metastore" script.
+        format!("bin/start-metastore --config {STACKABLE_CONFIG_DIR} --db-type {db_type} --hive-bin-dir bin &")
+    } else {
+        // schematool versions 4.0.x (and above) support the `-initOrUpgradeSchema`, which is exactly what we need :)
+        // Some docs for the schemaTool can be found here: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=34835119
+        formatdoc! {"
+            bin/base --config {STACKABLE_CONFIG_DIR} --service schemaTool -dbType {db_type} -initOrUpgradeSchema
+            bin/base --config {STACKABLE_CONFIG_DIR} --service metastore &
+        "}
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/controller.rs]: Makes sense, did so in 68ded98. I did explicitly choose \"{foo}\" oder {foo:?}

[the remark this answers is not in the document]
Comment (NickLarsenNZ) [reply rust/operator-binary/src/controller.rs]: There were more occurrences, but that's fine they can be done over time.

**Comment by sbernauer** [comment]
Started https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hive-operator-it-custom/6

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM

**Comment by sbernauer** [comment]
https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hive-operator-it-custom/7/

**Comment by sbernauer** [comment]
All tests passed

**Pull request merged**